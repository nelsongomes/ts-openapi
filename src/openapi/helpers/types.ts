import Joi from "joi";

type Common = {
  description?: string;
  required?: boolean;
  nullable?: boolean;
};

type ParameterCommon = { isParameter?: boolean };
type ObjectCommon = { modelName?: string };

type StringParameters = {
  default?: string;
  example?: string;
  minLength?: number;
  maxLength?: number;
} & Common &
  ParameterCommon;

type StringEnumParameters = {
  values: string[];
  default?: string;
  example?: string;
} & Common &
  ParameterCommon;

type EmailParameters = StringParameters;
type UriParameters = StringParameters;
type HostnameParameters = StringParameters;
type PasswordParameters = {
  minLength?: number;
  maxLength?: number;
} & Common &
  ParameterCommon;
type UuidParameters = {
  example?: string;
} & Common &
  ParameterCommon;
type IpParameters = {
  example?: string;
} & Common &
  ParameterCommon;
type BinaryParameters = StringParameters;
type ByteParameters = StringParameters;

type DateParameters = {
  default?: string;
  example?: string;
} & Common &
  ParameterCommon;

type IntegerParameters = {
  default?: number;
  example?: number;
  minValue?: number;
  maxValue?: number;
} & Common &
  ParameterCommon;

type IntegerEnumParameters = {
  values: any[];
  default?: number;
  example?: number;
} & Common &
  ParameterCommon;

type NumberEnumParameters = IntegerEnumParameters & ParameterCommon;

type BooleanParameters = {
  default?: boolean;
  example?: boolean;
} & Common &
  ParameterCommon;

type ObjectParameters = {
  properties: Joi.SchemaMap<any>;
  default?: object;
  example?: object;
} & Common &
  ObjectCommon;

type ArrayParameters = {
  arrayType: Joi.SchemaLikeWithoutArray;
  default?: object;
  example?: object;
  minLength?: number;
  maxLength?: number;
} & Common &
  ParameterCommon;

export const Types = {
  String: (parameters?: StringParameters) => {
    let joiType = Joi.string();

    if (parameters) {
      const {
        description,
        required,
        nullable,
        maxLength,
        minLength,
        example
      } = parameters;

      if (description) {
        joiType = joiType.description(description);
      }

      if (typeof required === "boolean") {
        joiType = required ? joiType.required() : joiType.optional();
      }

      if (typeof nullable === "boolean" && nullable) {
        joiType = joiType.allow(null);
      }

      if (minLength) {
        joiType = joiType.min(minLength);
      }

      if (maxLength) {
        joiType = joiType.max(maxLength);
      }

      if (parameters.default) {
        joiType = joiType.default(parameters.default);
      }

      if (parameters.example) {
        joiType = joiType.example(example);
      }

      if (parameters.isParameter) {
        joiType = joiType.meta({ parameter: true });
      }
    }

    return joiType;
  },

  StringEnum: (parameters: StringEnumParameters) => {
    return Types.String(parameters).valid(...parameters.values);
  },

  Email: (parameters?: EmailParameters) => {
    return Types.String(parameters).email();
  },

  Password: (parameters?: PasswordParameters) => {
    return Types.String(parameters).meta({ format: "password" });
  },

  Uuid: (parameters?: UuidParameters) => {
    return Types.String(parameters)
      .uuid({ version: "uuidv4" })
      .meta({ format: "uuid" })
      .min(36)
      .max(36);
  },

  Uri: (parameters?: UriParameters) => {
    return Types.String(parameters)
      .uri()
      .meta({ format: "uri" });
  },

  Hostname: (parameters?: HostnameParameters) => {
    return Types.String(parameters)
      .hostname()
      .meta({ format: "hostname" });
  },

  Ipv4: (parameters?: IpParameters) => {
    return Types.String(parameters)
      .ip({ version: ["ipv4"] })
      .meta({ format: "ipv4" })
      .min(7)
      .max(15);
  },

  Ipv6: (parameters?: IpParameters) => {
    return Types.String(parameters)
      .ip({ version: ["ipv6"] })
      .meta({ format: "ipv6" })
      .min(3)
      .max(45);
  },

  Binary: (parameters?: BinaryParameters) => {
    let joiType = Joi.binary();

    if (parameters) {
      const {
        description,
        required,
        nullable,
        maxLength,
        minLength,
        example
      } = parameters;

      if (description) {
        joiType = joiType.description(description);
      }

      if (typeof required === "boolean") {
        joiType = required ? joiType.required() : joiType.optional();
      }

      if (typeof nullable === "boolean" && nullable) {
        joiType = joiType.allow(null);
      }

      if (minLength) {
        joiType = joiType.min(minLength);
      }

      if (maxLength) {
        joiType = joiType.max(maxLength);
      }

      if (parameters.default) {
        joiType = joiType.default(parameters.default);
      }

      if (parameters.example) {
        joiType = joiType.example(example);
      }

      if (parameters.isParameter) {
        joiType = joiType.meta({ parameter: true });
      }
    }

    return joiType;
  },

  Byte: (parameters?: ByteParameters) => {
    return Types.Binary(parameters).encoding("base64");
  },

  DateTime: (parameters?: DateParameters) => {
    let joiType = Joi.string()
      .isoDate()
      .max(24);

    if (parameters) {
      const { description, required, nullable } = parameters;

      if (description) {
        joiType = joiType.description(description);
      }

      if (typeof required === "boolean") {
        joiType = required ? joiType.required() : joiType.optional();
      }

      if (typeof nullable === "boolean" && nullable) {
        joiType = joiType.allow(null);
      }

      if (parameters.isParameter) {
        joiType = joiType.meta({ parameter: true });
      }
    }

    return joiType;
  },

  Date: (parameters?: DateParameters) => {
    return Types.DateTime(parameters)
      .meta({ format: "date" })
      .min(10)
      .max(10);
  },

  Number: (parameters?: IntegerParameters) => {
    let joiType = Joi.number();

    if (parameters) {
      const {
        description,
        required,
        nullable,
        minValue,
        maxValue,
        example
      } = parameters;

      if (description) {
        joiType = joiType.description(description);
      }

      if (typeof required === "boolean" && required) {
        joiType = joiType.required();
      }

      if (typeof nullable === "boolean" && nullable) {
        joiType = joiType.allow(null);
      }

      if (typeof minValue === "number") {
        joiType = joiType.min(minValue);
      }

      if (typeof maxValue === "number") {
        joiType = joiType.max(maxValue);
      }

      if (typeof parameters.default === "number") {
        joiType = joiType.default(parameters.default);
      }

      if (typeof parameters.example === "number") {
        joiType = joiType.example(example);
      }

      if (parameters.isParameter) {
        joiType = joiType.meta({ parameter: true });
      }
    }

    return joiType;
  },

  NumberEnum: (parameters: NumberEnumParameters) => {
    return Types.Number(parameters).valid(...parameters.values);
  },

  Integer: (parameters?: IntegerParameters) => {
    return Types.Number(parameters).integer();
  },

  IntegerEnum: (parameters: IntegerEnumParameters) => {
    return Types.Integer(parameters).valid(...parameters.values);
  },

  Boolean: (parameters?: BooleanParameters) => {
    let joiType = Joi.boolean().strict();

    if (parameters) {
      const { description, required, nullable, example } = parameters;

      if (description) {
        joiType = joiType.description(description);
      }

      if (typeof required === "boolean") {
        joiType = required ? joiType.required() : joiType.optional();
      }

      if (typeof nullable === "boolean" && nullable) {
        joiType = joiType.allow(null);
      }

      if (typeof parameters.default === "boolean") {
        joiType = joiType.default(parameters.default);
      }

      if (typeof parameters.example === "boolean") {
        joiType = joiType.example(example);
      }

      if (parameters.isParameter) {
        joiType = joiType.meta({ parameter: true });
      }
    }

    return joiType;
  },

  Object: (parameters: ObjectParameters) => {
    const { properties, description, required, nullable, example } = parameters;
    let joiType = Joi.object().keys(properties);

    if (description) {
      joiType = joiType.description(description);
    }

    if (typeof required === "boolean") {
      joiType = required ? joiType.required() : joiType.optional();
    }

    if (nullable) {
      joiType = joiType.allow(null);
    }

    if (parameters.default) {
      joiType = joiType.default(parameters.default);
    }

    if (parameters.example) {
      joiType = joiType.example(example);
    }

    if (parameters.modelName) {
      joiType = joiType.meta({ modelName: parameters.modelName });
    }

    return joiType;
  },

  Array: (parameters: ArrayParameters) => {
    const {
      arrayType,
      description,
      required,
      nullable,
      example,
      minLength,
      maxLength
    } = parameters;
    let joiType = Joi.array().items(arrayType);

    if (description) {
      joiType = joiType.description(description);
    }

    if (typeof required === "boolean") {
      joiType = required ? joiType.required() : joiType.optional();
    }

    if (nullable) {
      joiType = joiType.allow(null);
    }

    if (parameters.default) {
      joiType = joiType.default(parameters.default);
    }

    if (parameters.example) {
      joiType = joiType.example(example);
    }

    if (minLength) {
      joiType = joiType.min(minLength);
    }

    if (maxLength) {
      joiType = joiType.max(maxLength);
    }

    if (parameters.isParameter) {
      joiType = joiType.meta({ parameter: true });
    }

    return joiType;
  }
};
