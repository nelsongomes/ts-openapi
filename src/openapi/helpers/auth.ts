import {
  SecurityApiKeyScheme,
  SecurityApiKeySchemeMethod,
  SecurityBasicScheme,
  SecurityBearerScheme,
  SecurityOauth2Scheme,
  SecurityOauth2Scopes,
} from "../openapi.types";

export function basicAuth(): SecurityBasicScheme {
  return {
    type: "http",
    scheme: "basic",
  };
}

export function apiKeyAuth(
  name: string,
  method: SecurityApiKeySchemeMethod
): SecurityApiKeyScheme {
  return {
    type: "apiKey",
    in: method,
    name,
  };
}

export function cookieAuth(cookieName: string): SecurityApiKeyScheme {
  return {
    type: "apiKey",
    in: "cookie",
    name: cookieName,
  };
}

export function bearerAuth(
  bearerFormat?: "JWT" | string
): SecurityBearerScheme {
  return {
    type: "http",
    scheme: "bearer",
    ...(bearerFormat && { bearerFormat }),
  };
}

export function oauth2ImplicitAuth(
  description: string,
  authorizationUrl: string,
  scopes: SecurityOauth2Scopes,
  refreshUrl?: string
): SecurityOauth2Scheme {
  return {
    type: "oauth2",
    description,
    flows: {
      implicit: {
        authorizationUrl,
        scopes,
        refreshUrl,
      },
    },
  };
}

export function oauth2AuthorizationCodeAuth(
  description: string,
  authorizationUrl: string,
  tokenUrl: string,
  scopes: SecurityOauth2Scopes,
  refreshUrl?: string
): SecurityOauth2Scheme {
  return {
    type: "oauth2",
    description,
    flows: {
      authorizationCode: {
        authorizationUrl,
        tokenUrl,
        scopes,
        refreshUrl,
      },
    },
  };
}

export function oauth2PasswordAuth(
  description: string,
  tokenUrl: string,
  scopes: SecurityOauth2Scopes,
  refreshUrl?: string
): SecurityOauth2Scheme {
  return {
    type: "oauth2",
    description,
    flows: {
      password: {
        tokenUrl,
        scopes,
        refreshUrl,
      },
    },
  };
}

export function oauth2ClientCredentialsAuth(
  description: string,
  tokenUrl: string,
  scopes: SecurityOauth2Scopes,
  refreshUrl?: string
): SecurityOauth2Scheme {
  return {
    type: "oauth2",
    description,
    flows: {
      clientCredentials: {
        tokenUrl,
        scopes,
        refreshUrl,
      },
    },
  };
}
