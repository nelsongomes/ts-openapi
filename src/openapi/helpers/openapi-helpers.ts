import {
  Body,
  SecurityApiKeyScheme,
  SecurityApiKeySchemeMethod,
  SecurityBasicScheme,
  SecurityBearerScheme,
} from "../openapi.types";

export function textPlain(description: string): Body {
  return {
    description,
    content: { "text-plain": {} },
  };
}

export function basicAuth(): SecurityBasicScheme {
  return {
    type: "http",
    scheme: "basic",
  };
}

export function apiKeyAuth(
  name: string,
  method: SecurityApiKeySchemeMethod
): SecurityApiKeyScheme {
  return {
    type: "apiKey",
    in: method,
    name,
  };
}

export function cookieAuth(cookieName: string): SecurityApiKeyScheme {
  return {
    type: "apiKey",
    in: "cookie",
    name: cookieName,
  };
}

export function bearerAuth(
  bearerFormat?: "JWT" | string
): SecurityBearerScheme {
  return {
    type: "http",
    scheme: "bearer",
    ...(bearerFormat && { bearerFormat }),
  };
}
