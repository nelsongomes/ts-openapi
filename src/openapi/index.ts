import Joi, { Schema } from "joi";

export { Joi, Schema };

export { OpenApi } from "./openapi";
export { WebRequestSchema, ParameterIn, Parameters } from "./openapi.types";

export { textPlain } from "./openapi-helpers";
