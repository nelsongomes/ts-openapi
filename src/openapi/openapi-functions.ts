import Joi, { ObjectSchema } from "joi";
import joiToSwagger from "../joi-conversion";
import {
  Body,
  IntegerFormats,
  NumberFormats,
  SchemaTypeArray,
  SchemaTypeBoolean,
  SchemaTypeInteger,
  SchemaTypeNumber,
  SchemaTypeObject,
  SchemaTypeString,
  StringFormats
} from "./openapi.types";

export function limitations(parameter: any): string {
  const limitationArray = [];

  if (parameter.type === "string" && parameter.enum) {
    // min & max should be ignored when is an enum
    delete parameter.minLength;
    delete parameter.maxLength;
  }

  // string
  if (parameter.minLength) {
    limitationArray.push(`min:${parameter.minLength} chars`);
  }

  if (parameter.maxLength) {
    limitationArray.push(`max:${parameter.maxLength} chars`);
  }

  // integer / float
  if (parameter.type === "integer" && parameter.enum) {
    // min & max should be ignored when is an enum
    delete parameter.minimum;
    delete parameter.maximum;
  }

  if (typeof parameter.minimum === "number") {
    limitationArray.push(`min:${parameter.minimum}`);
  }

  if (typeof parameter.maximum === "number") {
    limitationArray.push(`max:${parameter.maximum}`);
  }

  // array
  if (parameter.minItems) {
    limitationArray.push(`minItems:${parameter.minItems}`);
  }

  if (parameter.maxItems) {
    limitationArray.push(`maxItems:${parameter.maxItems}`);
  }

  if (parameter.format) {
    switch (parameter.format) {
      case StringFormats.DateTime:
        if (parameter.meta && parameter.meta.format === "date") {
          limitationArray.push(`date:yyyy-mm-dd`);
        } else {
          limitationArray.push(`ISO8601 date-time format`);
        }
        break;
      case StringFormats.Byte:
        limitationArray.push("base64 encoded string");
        break;
      case StringFormats.Binary:
        limitationArray.push("binary string");
        break;
    }
  }

  return limitationArray.length > 0 ? ` (${limitationArray.join(", ")})` : "";
}

export function bodyParams(schema: ObjectSchema): Body {
  const internalSchema = Joi.object().keys({ object: schema.required() });
  const query = joiToSwagger(internalSchema, {});

  const key = Object.keys(query.swagger.properties)[0];
  const parameter = query.swagger.properties[key];

  return {
    description:
      query.swagger.properties[key].description ||
      "Body does not have a description.",
    content: {
      "application/json": {
        schema: objectSchema(parameter),
        ...(parameter.example && { example: parameter.example })
      }
    }
  };
}

function objectSchema(parameter: any): SchemaTypeObject {
  const description =
    (parameter.description || "Parameter without description.") +
    (limitations(parameter) || "");

  const output: SchemaTypeObject = {
    description,
    ...(typeof parameter.default === "object" && {
      default: parameter.default
    }),
    ...(parameter.minItems && { minItems: parameter.minItems }),
    ...(parameter.maxItems && { maxItems: parameter.maxItems }),
    ...(typeof parameter.nullable === "boolean" && {
      nullable: parameter.nullable
    }),
    type: "object",
    properties: {}
  };

  for (const propertyKey of Object.keys(parameter.properties)) {
    const property = parameter.properties[propertyKey];

    switch (property.type) {
      case "string":
        const {
          default: ignoreA1,
          nullable: ignoreA2,
          minLength,
          maxLength,
          ...restOfStringSchema
        } = stringSchema(property);

        output.properties[propertyKey] = restOfStringSchema;
        break;
      case "number":
        const {
          default: ignoreB1,
          nullable: ignoreB2,
          minimum: ignoreB3,
          maximum: ignoreB4,
          ...restNumberOfSchema
        } = numberSchema(property);

        output.properties[propertyKey] = restNumberOfSchema;
        break;
      case "boolean":
        const {
          default: ignoreE1,
          nullable: ignorE2,
          ...restBooleanOfSchema
        } = booleanSchema(property);

        output.properties[propertyKey] = restBooleanOfSchema;
        break;
      case "integer":
        const {
          default: ignoreC1,
          nullable: ignoreC2,
          minimum: ignoreC3,
          maximum: ignoreC4,
          ...restIntegerOfSchema
        } = integerSchema(property);

        output.properties[propertyKey] = restIntegerOfSchema;
        break;
      case "object":
        const {
          default: ignoreD1,
          nullable: ignoreD2,
          ...restOfObjectSchema
        } = objectSchema(property);

        output.properties[propertyKey] = restOfObjectSchema;
        break;
      case "array":
        const {
          default: ignoreF1,
          nullable: ignoreF2,
          ...restOfArraySchema
        } = arraySchema(property);

        output.properties[propertyKey] = restOfArraySchema;
        break;
      default:
        throw new Error(`${property.type} not implemented`);
    }
  }

  return output;
}

export function stringSchema(parameter: any): SchemaTypeString {
  const supportedFormats = Object.values(StringFormats);
  const description =
    (parameter.description || "Parameter without description.") +
    (limitations(parameter) || "");

  const output: SchemaTypeString = {
    description,
    ...(parameter.default && { default: parameter.default }),
    ...(parameter.format &&
      supportedFormats.includes(parameter.format) && {
        format: parameter.format
      }),
    ...(parameter.enum && { enum: parameter.enum }),
    ...(parameter.minLength && { minLength: parameter.minLength }),
    ...(parameter.maxLength && { maxLength: parameter.maxLength }),
    ...(typeof parameter.nullable === "boolean" && {
      nullable: parameter.nullable
    }),
    type: "string"
  };

  if (parameter.meta) {
    if (parameter.meta.format) {
      parameter.format = parameter.meta.format;
    }
  }

  if (parameter.format && supportedFormats.includes(parameter.format)) {
    output.format = parameter.format;

    switch (parameter.format) {
      case StringFormats.Date:
        output.maxLength = 10; // 2020-10-14
        output.minLength = 10;
        break;
      case StringFormats.DateTime:
        output.minLength = 16; // 20201014T214403Z
        output.maxLength = 25; // 2020-10-14T21:44:03+00:00
        break;
      case StringFormats.Password:
        delete output.default; // no defaults for passwords
        break;
    }
  }

  // default values must be part of enum
  if (
    parameter.enum instanceof Array &&
    !parameter.enum.includes(output.default)
  ) {
    delete output.default;
  }

  return output;
}

export function numberSchema(parameter: any): SchemaTypeNumber {
  const description =
    (parameter.description || "Parameter without description.") +
    (limitations(parameter) || "");

  const output: SchemaTypeNumber = {
    description,
    ...(parameter.default && { default: parameter.default }),
    ...(parameter.enum && { enum: parameter.enum }),
    ...(typeof parameter.minimum === "number" && {
      minimum: parameter.minimum
    }),
    ...(typeof parameter.maximum === "number" && {
      maximum: parameter.maximum
    }),
    ...(typeof parameter.nullable === "boolean" && {
      nullable: parameter.nullable
    }),
    format: NumberFormats.Double,
    type: "number"
  };

  return output;
}

export function booleanSchema(parameter: any): SchemaTypeBoolean {
  const description =
    (parameter.description || "Parameter without description.") +
    (limitations(parameter) || "");

  const output: SchemaTypeBoolean = {
    description,
    ...(typeof parameter.default === "boolean" && {
      default: parameter.default
    }),
    ...(typeof parameter.nullable === "boolean" && {
      nullable: parameter.nullable
    }),
    type: "boolean"
  };

  return output;
}

export function integerSchema(parameter: any): SchemaTypeInteger {
  const description =
    (parameter.description || "Parameter without description.") +
    (limitations(parameter) || "");

  const output: SchemaTypeInteger = {
    description,
    ...(typeof parameter.default === "number" && {
      default: parameter.default
    }),
    ...(typeof parameter.example === "number" && {
      example: parameter.example
    }),
    ...(parameter.enum && { enum: parameter.enum }),
    ...(typeof parameter.minimum === "number" && {
      minimum: parameter.minimum
    }),
    ...(typeof parameter.maximum === "number" && {
      maximum: parameter.maximum
    }),
    ...(typeof parameter.nullable === "boolean" && {
      nullable: parameter.nullable
    }),
    format: IntegerFormats.Int64,
    type: "integer"
  };

  // default values must be part of enum
  if (
    parameter.enum instanceof Array &&
    !parameter.enum.includes(output.default)
  ) {
    delete output.default;
  }

  return output;
}

export function arraySchema(parameter: any): SchemaTypeArray {
  const output: SchemaTypeArray = {
    ...(typeof parameter.default === "object" && {
      default: parameter.default
    }),
    ...(parameter.minItems && { minItems: parameter.minItems }),
    ...(parameter.maxItems && { maxItems: parameter.maxItems }),
    ...(typeof parameter.nullable === "boolean" && {
      nullable: parameter.nullable
    }),
    type: "array"
  };

  switch (parameter.items.type) {
    case "string":
      const {
        default: ignoreA1,
        nullable: ignoreA2,
        minLength,
        maxLength,
        ...restOfStringSchema
      } = stringSchema(parameter.items);

      output.items = restOfStringSchema;
      break;
    case "number":
      const {
        default: ignoreB1,
        nullable: ignoreB2,
        minimum: ignoreB3,
        maximum: ignoreB4,
        ...restNumberOfSchema
      } = numberSchema(parameter.items);

      output.items = restNumberOfSchema;
      break;
    case "integer":
      const {
        default: ignoreC1,
        nullable: ignoreC2,
        minimum: ignoreC3,
        maximum: ignoreC4,
        ...restIntegerOfSchema
      } = integerSchema(parameter.items);

      output.items = restIntegerOfSchema;
      break;
    case "object":
      const {
        default: ignoreD1,
        nullable: ignoreD2,
        ...restOfObjectSchema
      } = objectSchema(parameter.items);

      output.items = restOfObjectSchema;
      break;
    default:
      throw new Error("not implemented");
  }

  if (parameter.items.enum instanceof Array) {
    // default values must be part of enum
    if (
      output.default instanceof Array &&
      !output.default.every((element: any, _index: number, _array: any[]) =>
        parameter.items.enum.includes(element)
      )
    ) {
      delete output.default;
    }

    // string enums are already limited
    if (parameter.items.type === "string") {
      delete (output.items as SchemaTypeString).minLength;
      delete (output.items as SchemaTypeString).maxLength;
    }
  }

  return output;
}

export function bodySchema(schema: ObjectSchema): Body {
  return bodyParams(schema);
}
