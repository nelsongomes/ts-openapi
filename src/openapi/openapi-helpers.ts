export function textPlain(description: string) {
  return {
    content: {
      "text/plain": {
        schema: {
          type: "string",
        },
      },
    },
    description,
  };
}
