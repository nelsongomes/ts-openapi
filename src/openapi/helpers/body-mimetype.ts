import { Body } from "../openapi.types";

export function textPlain(description: string): Body {
  return {
    description,
    content: { "text-plain": {} },
  };
}
