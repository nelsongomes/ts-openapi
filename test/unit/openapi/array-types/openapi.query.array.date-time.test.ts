import Joi from "joi";
import { OpenApi } from "../../../../src/openapi/openapi";
import { textPlain } from "../../../../src/openapi/helpers/openapi-helpers";
import { Parameters, ParameterIn } from "../../../../src/openapi/openapi.types";

describe("src/openapi/openapi", () => {
  let openApi: OpenApi;

  describe("Array of date-time", () => {
    beforeEach(() => {
      openApi = new OpenApi(
        "1.0.0",
        "Server API",
        "Some test api",
        "nelson.ricardo.gomes@gmail.com"
      );

      openApi.setServers([{ url: "https://server.com" }]);
      openApi.addPath(
        "/health",
        {
          get: {
            description: "Service healthcheck endpoint",
            operationId: "repeated",
            parameters: [],
            responses: {
              200: textPlain("Successful operation."),
            },
            summary: "Server Healthcheck",
            tags: ["Internals"],
          },
        },
        true
      );
    });

    test("date-time array simple", async () => {
      const parameters: Parameters = [];

      const query = Joi.object().keys({
        timestamps: Joi.array().items(Joi.string().isoDate()),
      });

      openApi.genericParams(parameters, query, ParameterIn.Query);
      openApi.addPath(
        "/test",
        {
          get: {
            description: "Test endpoint",
            operationId: "id",
            parameters,
            responses: {
              200: textPlain("Successful operation."),
            },
            summary: "Server Test",
            tags: ["Internals"],
          },
        },
        true
      );
      expect(openApi.generateJson()).toMatchSnapshot();
    });

    test("date array simple", async () => {
      const parameters: Parameters = [];

      const query = Joi.object().keys({
        timestamps: Joi.array().items(
          Joi.string()
            .isoDate()
            .meta({ format: "date" })
        ),
      });

      openApi.genericParams(parameters, query, ParameterIn.Query);
      openApi.addPath(
        "/test",
        {
          get: {
            description: "Test endpoint",
            operationId: "id",
            parameters,
            responses: {
              200: textPlain("Successful operation."),
            },
            summary: "Server Test",
            tags: ["Internals"],
          },
        },
        true
      );
      expect(openApi.generateJson()).toMatchSnapshot();
    });

    test("date-time array all options", async () => {
      const parameters: Parameters = [];
      const query = Joi.object()
        .keys({
          timestamps: Joi.array()
            .items(
              Joi.string()
                .isoDate()
                // ignore these
                .description("ignore")
                .max(999)
                .allow(null)
                .required()
            )
            .min(1)
            .max(100)
            .required()
            .description("date-time array")
            .default(["2020-10-14T22:12:53.065Z", "2020-10-14T22:12:53.065Z"])
            .example(["2020-10-14T22:12:53.065Z"])
            .allow(null),
        })
        .description("ignore");

      openApi.genericParams(parameters, query, ParameterIn.Query);
      openApi.addPath(
        "/test",
        {
          get: {
            description: "Test endpoint",
            operationId: "id",
            parameters,
            responses: {
              200: textPlain("Successful operation."),
            },
            summary: "Server Test",
            tags: ["Internals"],
          },
        },
        true
      );
      expect(openApi.generateJson()).toMatchSnapshot();
    });
  });
});
