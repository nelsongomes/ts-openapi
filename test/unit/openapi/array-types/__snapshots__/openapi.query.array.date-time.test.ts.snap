// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/openapi/openapi Array of date-time date array simple 1`] = `
Object {
  "info": Object {
    "contact": Object {
      "email": "nelson.ricardo.gomes@gmail.com",
    },
    "description": "Some test api",
    "license": Object {
      "name": "Apache 2.0",
      "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "termsOfService": "http://swagger.io/terms/",
    "title": "Server API",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/health": Object {
      "get": Object {
        "description": "Service healthcheck endpoint",
        "operationId": "repeated",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Healthcheck",
        "tags": Array [
          "Internals",
        ],
      },
    },
    "/test": Object {
      "get": Object {
        "description": "Test endpoint",
        "operationId": "id",
        "parameters": Array [
          Object {
            "description": "Parameter without description.",
            "in": "query",
            "name": "timestamps",
            "required": false,
            "schema": Object {
              "items": Object {
                "format": "date",
                "type": "string",
              },
              "type": "array",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Test",
        "tags": Array [
          "Internals",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://server.com",
    },
  ],
}
`;

exports[`src/openapi/openapi Array of date-time date-time array all options 1`] = `
Object {
  "info": Object {
    "contact": Object {
      "email": "nelson.ricardo.gomes@gmail.com",
    },
    "description": "Some test api",
    "license": Object {
      "name": "Apache 2.0",
      "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "termsOfService": "http://swagger.io/terms/",
    "title": "Server API",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/health": Object {
      "get": Object {
        "description": "Service healthcheck endpoint",
        "operationId": "repeated",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Healthcheck",
        "tags": Array [
          "Internals",
        ],
      },
    },
    "/test": Object {
      "get": Object {
        "description": "Test endpoint",
        "operationId": "id",
        "parameters": Array [
          Object {
            "description": "date-time array (minItems:1, maxItems:100)",
            "example": Array [
              "2020-10-14T22:12:53.065Z",
            ],
            "in": "query",
            "name": "timestamps",
            "required": true,
            "schema": Object {
              "default": Array [
                "2020-10-14T22:12:53.065Z",
                "2020-10-14T22:12:53.065Z",
              ],
              "items": Object {
                "format": "date-time",
                "type": "string",
              },
              "maxItems": 100,
              "minItems": 1,
              "nullable": true,
              "type": "array",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Test",
        "tags": Array [
          "Internals",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://server.com",
    },
  ],
}
`;

exports[`src/openapi/openapi Array of date-time date-time array simple 1`] = `
Object {
  "info": Object {
    "contact": Object {
      "email": "nelson.ricardo.gomes@gmail.com",
    },
    "description": "Some test api",
    "license": Object {
      "name": "Apache 2.0",
      "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "termsOfService": "http://swagger.io/terms/",
    "title": "Server API",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/health": Object {
      "get": Object {
        "description": "Service healthcheck endpoint",
        "operationId": "repeated",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Healthcheck",
        "tags": Array [
          "Internals",
        ],
      },
    },
    "/test": Object {
      "get": Object {
        "description": "Test endpoint",
        "operationId": "id",
        "parameters": Array [
          Object {
            "description": "Parameter without description.",
            "in": "query",
            "name": "timestamps",
            "required": false,
            "schema": Object {
              "items": Object {
                "format": "date-time",
                "type": "string",
              },
              "type": "array",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Test",
        "tags": Array [
          "Internals",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://server.com",
    },
  ],
}
`;
